### ISSUE TITLE
[Bug]: CPU RAM OOM when doing GPTQ on 0.6B model

### ISSUE THREAD

**Issue by gau-nernst** [vllm-project/llm-compressor#1754]
### ⚙️ Your current environment

<details>
<summary>The output of <code>python collect_env.py</code></summary>

```text
### Environment Information ###
Operating System: `Linux-6.8.0-60-generic-x86_64-with-glibc2.39`
Python Version: `3.12.10 (main, May 30 2025, 05:37:37) [Clang 20.1.4 ]`
llm-compressor Version: `0.6.0`
compressed-tensors Version: `0.10.2`
transformers Version: `4.55.2`
torch Version: `2.8.0`
CUDA Devices: `['NVIDIA RTX A6000', 'NVIDIA RTX A6000', 'NVIDIA RTX A6000', 'NVIDIA RTX A6000', 'NVIDIA RTX A6000', 'NVIDIA RTX A6000', 'NVIDIA RTX A6000', 'NVIDIA RTX A6000']`
AMD Devices: `None`
```
</details>

### 🐛 Describe the bug

```python
# debug.py
from datasets import load_dataset
from llmcompressor import oneshot
from llmcompressor.modifiers.quantization import GPTQModifier
from transformers import AutoTokenizer


def main():
    model_id = "Qwen/Qwen3-0.6B"
    save_dir = "out/Qwen3-0.6B"
    ds_name = "neuralmagic/LLM_compression_calibration"
    num_samples = 512
    max_seq_len = 8192

    tokenizer = AutoTokenizer.from_pretrained(model_id)

    def preprocess_fn(example):
        return {"text": tokenizer.apply_chat_template(example["messages"], add_generation_prompt=False, tokenize=False)}

    ds = load_dataset(ds_name, split="train")
    ds = ds.shuffle(seed=2024).select(range(num_samples))  # random sample
    ds = ds.map(preprocess_fn)

    recipe = GPTQModifier(ignore=["lm_head"], scheme="W4A16")

    oneshot(
        model=model_id,
        dataset=ds,
        recipe=recipe,
        max_seq_length=max_seq_len,
        output_dir=save_dir,
        num_calibration_samples=num_samples,
    )


if __name__ == "__main__":
    main()
```
```
CUDA_VISIBLE_DEVICES=0 python debug.py
```

When running the above snippet, during "Propagating" phase, I observe that RAM usage will slowly increase until it OOM the machine. It looks like some kind of memory leak / data not free.

I saw some related issues like #1650 but not sure if it's the same problem.

**Comment by gau-nernst** [comment]
I have more time to debug today and have identified the issue
- The root cause is `F.sdpa()` falls back to naive, slow, math implementation, which materializes the attention matrix -> quadratic of sequence length.
- The fallback is caused by LLM Compressor passing `attn_mask` to `F.sdpa()`. 
- The fix is simple: force FA2 impl from `flash-attn` package instead, which has support for `attn_mask`

```python
from transformers import AutoModelForCausalLM

model = AutoModelForCausalLM.from_pretrained(
    model_id,
    torch_dtype="auto",
    attn_implementation="flash_attention_2",
)

...

oneshot(model, ...)
```

There are 2 follow up questions
1. Is `F.sdpa()` falling back to math implementation when `attn_mask` is used expected? This is the question for the PyTorch team, so not relevant for LLM Compressor
2. It seems like LLM Compressor does forward pass for 1 sample at a time (from my loggings of F.sdpa(), not 100% sure for all cases). It means that using `attn_mask` is actually unnecessary. Can LLM Compressor remove this attention mask so that `F.sdpa()` can use flash attention impl out of the box?

The second question is more like an open question for the dev team. I'm closing this issue now since the root cause is not from LLM Compressor, and I have a workaround for it. Though it would be best if this can be documented in an example as well.

Update: turns out it's not that `flash-attn` supports attention mask, but HF has special logic to do sequence packing for `flash-attn`

https://github.com/huggingface/transformers/blob/ff8b88a948fc2f6aba421ca64ad165291928dcee/src/transformers/modeling_flash_attention_utils.py#L634-L637
